alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java

```diff
@@ -90,37 +91,70 @@ public LoadConstants(TagConfig config) throws Exception {
 	}
 
 	@Override
-	public void apply(FaceletContext facesContext, UIComponent parentUIComponent) throws IOException {
+	public void apply(FaceletContext facletContext, UIComponent parentUIComponent) throws IOException {
+
+		Cache<String, Map<String, Object>> constantCache = LoadConstants.constantCache;
+
+		// First check without locking (not yet thread-safe)
+		if (constantCache == null) {
+
+			// Second check with locking (thread-safe)
+			synchronized (LoadConstants.class) {
+
+				constantCache = LoadConstants.constantCache;
+
+				if (constantCache == null) {
+
+					FacesContext facesContext = facletContext.getFacesContext();
+					ExternalContext externalContext = facesContext.getExternalContext();
+					String maxCacheCapacityString = externalContext.getInitParameter(LoadConstants.class.getName() +
+							".maxCacheCapacity");
+
+					if (maxCacheCapacityString != null) {
+
+						CacheFactory cacheFactory = (CacheFactory) FactoryExtensionFinder.getFactory(externalContext,
+								CacheFactory.class);
+						int initialCacheCapacity = cacheFactory.getDefaultInitialCapacity();
+						int maxCacheCapacity = Integer.parseInt(maxCacheCapacityString);
+						constantCache = LoadConstants.constantCache = cacheFactory.getConcurrentCache(
+									initialCacheCapacity, maxCacheCapacity);
+					}
+					else {
+						constantCache = LoadConstants.constantCache = CacheFactory.getConcurrentCacheInstance(
+									externalContext);
+					}
+				}
+			}
+		}
 
-		if (cacheable && CONSTANT_MAP_CACHE.containsKey(classType)) {
-			facesContext.setAttribute(var, CONSTANT_MAP_CACHE.get(classType));
+		if (cacheable && constantCache.containsKey(classType)) {
+			facletContext.setAttribute(var, constantCache.get(classType));
 		}
 		else {
 
 			Map<String, Object> constantMap = new HashMap<String, Object>();
 
 			try {
+
 				Class<?> clazz = Class.forName(classType);
 				Field[] fields = clazz.getFields();
 
 				for (Field field : fields) {
+
 					int modifiers = field.getModifiers();
 
-					if (Modifier.isStatic(modifiers)) {
+					if (Modifier.isPublic(modifiers) && Modifier.isStatic(modifiers) && Modifier.isFinal(modifiers)) {
```

Comment:
@stiemannkj1 Restricting to `public` should be OK, but restricting to `final` could cause a regression in downstream developer usage of `alloy:loadConstants` if they neglected to add the keyword in their code.